python/python/tests/test_pubsub.py

```diff
@@ -0,0 +1,486 @@
+from __future__ import annotations
+import pytest
+from typing import List, Set, Dict, Tuple, Optional, Any, Awaitable
+import asyncio
+import fnmatch
+
+from tests.conftest import create_client
+
+from glide.async_commands.core import (
+    CoreCommands,
+)
+
+
+MessagesByExactChannel = Dict[str, Dict[str, int]] # channel -> {msg -> cnt}
+MessagesByChannelAndChannel = Dict[Tuple[str, Optional[str]], Dict[str, int]] # (channel, pattern) -> {msg -> cnt}
+SubscriptionsByMode = Dict[CoreCommands.ChannelModes, Set[str]] # subscription_type -> {channel_or_pattern}
+
+
+@pytest.mark.asyncio
+class TestPubSub:
+
+    class PubSubClient:
+        def __init__(self,
+                     subsciptions: SubscriptionsByMode,
+                     sharded_messages_to_publish: MessagesByExactChannel,
+                     unsharded_messages_to_publish: MessagesByExactChannel,
+                     cluster_mode: bool,
+                     request):
+            self.messages_received: MessagesByChannelAndChannel = {}
+            self.messages_to_receive: MessagesByChannelAndChannel = {}
+
+            self.sharded_messages_to_publish: MessagesByExactChannel = sharded_messages_to_publish
+            self.unsharded_messages_to_publish: MessagesByExactChannel = unsharded_messages_to_publish
+
+            self.subsciptions = subsciptions
+
+            self.cluster_mode = cluster_mode
+            self.request = request
+
+            self.redis_client: CoreCommands = None
+
+
+        @staticmethod
+        def handle_new_message(message: CoreCommands.PubSubMsg, client: TestPubSub.PubSubClient) -> None:
+            key = (message.channel, message.pattern)
+            if key not in client.messages_received:
+                    client.messages_received[key] = {}
+
+            if message not in client.messages_received[key]:
+                    client.messages_received[key][message] = 0
+
+            client.messages_received[key][message] += 1
+
+
+        async def connect_and_subscribe(self) -> List[Any]:
+            self.redis_client = await create_client(self.request, self.cluster_mode)
+
+            for channel_mode, channels_or_patterns in self.subsciptions:
+                await self.redis_client.subscribe(channels_or_patterns=channels_or_patterns,
+                                                  channel_mode=channel_mode,
+                                                  callback=TestPubSub.PubSubClient.handle_new_message,
+                                                  context=self)
+
+        def assert_client(self) -> None:
```

Comment:
assert_client  => verify_received_expected_messages

Reply:
rejecting, assert_client is more suitable in my opinion

Reply:
`assert_client` - what? the name is doesn't explain what the function does.

Reply:
it asserts client. Explanatory enough, done